lib/Dialect/SV/Transforms/HWRaiseInOutPorts.cpp

```diff
@@ -0,0 +1,195 @@
+//===- HWRaiseInOutPorts.cpp - Generator Callout Pass ---------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Call arbitrary programs and pass them the attributes attached to external
+// modules.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetail.h"
+#include "circt/Dialect/HW/HWInstanceGraph.h"
+#include "circt/Dialect/HW/HWOps.h"
+#include "circt/Dialect/HW/PortConverter.h"
+#include "circt/Dialect/SV/SVPasses.h"
+#include "mlir/IR/Builders.h"
+#include "llvm/ADT/PostOrderIterator.h"
+
+using namespace circt;
+using namespace sv;
+using namespace hw;
+
+//===----------------------------------------------------------------------===//
+// HWRaiseInOutPortsPass
+//===----------------------------------------------------------------------===//
+
+namespace {
+
+struct HWRaiseInOutPortsPass
+    : public sv::HWRaiseInOutPortsBase<HWRaiseInOutPortsPass> {
+  void runOnOperation() override;
+};
+} // end anonymous namespace
+
+namespace {
+
+class HWInOutPortConversion : public PortConversion {
+public:
+  HWInOutPortConversion(PortConverterImpl &converter, hw::PortInfo port);
+
+  void mapInputSignals(OpBuilder &b, Operation *inst, Value instValue,
+                       SmallVectorImpl<Value> &newOperands,
+                       ArrayRef<Backedge> newResults) override;
+  void mapOutputSignals(OpBuilder &b, Operation *inst, Value instValue,
+                        SmallVectorImpl<Value> &newOperands,
+                        ArrayRef<Backedge> newResults) override;
+
+private:
+  void buildInputSignals() override;
+  void buildOutputSignals() override;
+
+  // Readers of this port internal in the module.
+  llvm::SmallVector<sv::ReadInOutOp, 4> readers;
+  // Writers of this port internal in the module.
+  llvm::SmallVector<sv::AssignOp, 4> writers;
+
+  bool hasReaders() { return !readers.empty(); }
+  bool hasWriters() { return !writers.empty(); }
+
+  // Handles to port info of the newly created ports.
+  PortInfo readPort, writePort;
+};
+
+HWInOutPortConversion::HWInOutPortConversion(PortConverterImpl &converter,
+                                             hw::PortInfo port)
+    : PortConversion(converter, port) {
+  // Gather readers and writers (how to handle sv.passign?)
+  for (auto *user : body->getArgument(port.argNum).getUsers()) {
+    if (auto read = dyn_cast<sv::ReadInOutOp>(user))
+      readers.push_back(read);
+    else if (auto write = dyn_cast<sv::AssignOp>(user))
+      writers.push_back(write);
+    else
+      user->emitWarning() << "Use of inout port " << port.name
+                          << " is not supported";
+  }
+
+  if (writers.size() > 1)
+    converter.getModule()->emitWarning()
```

Comment:
I think this should be an error rather than a warning (and it should fail the pass). Hardware designers tend to ignore warnings, so I tend to prefer avoiding warnings, especially when they have this big of an effect on design correctness.

Reply:
Agreed. Modified the infra slightly to allow `PortConversion`s to provide a failable `init` function.